mlir/lib/Target/IRDLToCpp/IRDLToCpp.cpp

```diff
@@ -60,6 +61,35 @@ static std::string joinNameList(llvm::ArrayRef<std::string> names) {
   return nameArray;
 }
 
+static llvm::LogicalResult isSnakeCase(llvm::StringRef in,
+                                       mlir::Operation *loc) {
+  bool allowUnderscore = false;
+  for (auto &elem : in) {
+    if (elem == '_') {
+      if (!allowUnderscore)
+        return loc->emitError(
+            llvm::formatv("Error in symbol name `{}`: No leading or double "
+                          "underscores allowed.",
+                          in));
+    } else {
+      if (!isalnum(elem))
+        return loc->emitError(
+            llvm::formatv("Error in symbol name `{}`: Only numbers and "
+                          "lower-case characters allowed.",
+                          in));
+
+      if (llvm::isUpper(elem))
+        return loc->emitError(llvm::formatv(
+            "Error in symbol name `{}`: Upper-case characters are not allowed.",
+            in));
+    }
+
+    allowUnderscore = elem != '_';
+  }
```

Comment:
I can't easily format on GitHub, consider committing it yourself.
```suggestion
  for (auto &elem : in) {
    if (elem == '_') {
      if (!allowUnderscore) {
        return loc->emitError(
            llvm::formatv("Error in symbol name `{}`: No leading or double "
                          "underscores allowed.",
                          in));
       }
       allowUnderscore = false;
       continue;
    }
    
      if (!isalnum(elem))
        return loc->emitError(
            llvm::formatv("Error in symbol name `{}`: Only numbers and "
                          "lower-case characters allowed.",
                          in));

      if (llvm::isUpper(elem))
        return loc->emitError(llvm::formatv(
            "Error in symbol name `{}`: Upper-case characters are not allowed.",
            in));

    allowUnderscore = true;
  }
```

Reply:
Also please use streams not formatv.

Reply:
Ah snap this was on older code. Let me move it.